Tests/Common/SimpleTcpListener.cs

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#nullable enable
+
+namespace LoRaWan.Tests.Common
+{
+    using System;
+    using System.Net;
+    using System.Net.Sockets;
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    public sealed class SimpleTcpListener : IDisposable
+    {
+        public sealed class Context
+        {
+            private readonly TcpClient client;
+
+            public Context(TcpClient client) => this.client = client;
+
+            public EndPoint RemoteEndPoint => this.client.Client.RemoteEndPoint;
+            public EndPoint LocalEndPoint => this.client.Client.LocalEndPoint;
+        }
+
+        private TcpListener? listener;
+
+        private SimpleTcpListener(TcpListener listener) =>
+            this.listener = listener;
+
+        public static SimpleTcpListener Start(int port, Func<Context, NetworkStream, Task> processor) =>
+            Start(port, null, processor);
+
+        public static SimpleTcpListener Start(int port, int? backlog, Func<Context, NetworkStream, Task> processor)
+        {
+            var listener = TcpListener.Create(port);
+            if (backlog is { } someBackLog)
+                listener.Start(someBackLog);
+            else
+                listener.Start();
+            _ = ListenAsync(listener, processor);
+            return new SimpleTcpListener(listener);
+        }
+
+        public void Dispose()
+        {
+            var listener = this.listener;
+            if (listener is null || listener != Interlocked.CompareExchange(ref this.listener, null, listener))
+                return;
+            listener.Stop();
+        }
+
+        private static async Task ListenAsync(TcpListener listener, Func<Context, NetworkStream, Task> processor)
+        {
+            while (true)
+                _ = OnAcceptAsync(await listener.AcceptTcpClientAsync().ConfigureAwait(false));
```

Comment:
Why do we not await this? If an error happens in there it goes unobserved.

Reply:
Because we don't want the processing of one client connection to block the accepting of another. That said, there is a chance that `AcceptTcpClientAsync` or `OnAcceptAsync` (up to its first await) could have thrown so I've revised this for logging and resilience here with 12db0c112a6485b89f02e35f602e04323a64de4a. I've also renamed `OnAcceptAsync` to `OnProcessAsync` with d208e5748a3b84cbd9ec4ba0b8db90c6d30321a4 to the accepting and processing more distinct.